dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1022 @@
+}
+
+public record CompositePayloadCodec(IEnumerable<PayloadCodec> Codecs) : IPayloadCodec
+{
+    public IEnumerable<Temporal.Api.Common.V1.Payload> Decode(IReadOnlyCollection<Temporal.Api.Common.V1.Payload> payloads) { /*...*/ }
+    
+    public IEnumerable<Temporal.Api.Common.V1.Payload> Encode(IReadOnlyCollection<Temporal.Api.Common.V1.Payload> payloads) { /*...*/ }
+}
+```
+
+Notes:
+
+* Should we prepare for some kind of .NET sandbox now by changing `DataConverter`'s `IPayloadConverter PayloadConverter`
+  to instead be a `Type PayloadConverterType`? (don't think we want generics there though)
+* Each top-level class/interface is of course its own file. Technically the nested classes violate some .NET
+  conventions (https://learn.microsoft.com/en-us/dotnet/standard/design-guidelines/nested-types), but those are more
+  general guidelines, and having a top-level file for every one of these things is annoying
+
+## Exceptions
+
+Looks like:
+
+```csharp
+namespace Temporal.Exceptions;
+
+public abstract class TemporalException : Exception
+{
+    internal TemporalException(string? message = null, Exception? inner = null) : base(message, inner) { /*...*/ }
+}
+
+public class FailureException : TemporalException
+{
+    internal FailureException(
+        string? message = null,
+        Exception? inner = null,
+        Temporal.Api.Failure.V1.Failure? failure = null) : base(message, inner) { /*...*/ }
+
+    public Temporal.Api.Failure.V1.Failure? Failure { get; private init; }
+}
+
+// Intentionally extensible
+public class AppException : FailureException
+{
+
+    public AppException(IEnumerable<object>? details = null, string? type = null, bool nonRetryable = false)
+    {
+        Type = type;
+        NonRetryable = nonRetryable;
+    }
+
+    public string? Type { get; protected init; }
+
+    public bool NonRetryable { get; protected init; }
+
+    public IEnumerable<object>? OutboundDetails { get; protected init; }
+
+    public IReadOnlyCollection<Temporal.Api.Common.V1.Payload>? InboundDetails { get; protected init; }
+
+    public bool HasDetails { get { /*...*/ } }
+}
+
+// Many more not listed
+```
+
+Notes:
+* Several more not listed. We will also include non-failure exceptions (e.g. `WorkflowAlreadyStartedException`) here
+  instead of in the client area. This namespace is the .NET equivalent of
+  https://pkg.go.dev/go.temporal.io/api/serviceerror + https://pkg.go.dev/go.temporal.io/sdk/temporal wrt errors.
+* `AppException` is not `ApplicationException` to avoid clash with `System.ApplicationException`
+* `AppException` is the only inheritable exception
+* We have to differentiate outbound (converted) details and inbound (not converted) details since, like some other SDKs,
+  we cannot convert this stuff until the type is provided to do so
+  * Should we store the converter on the exception for easy conversion? Then we could have
+  `public T GetDetail<T>(int index = 0)`
```

Comment:
I think that makes sense and is safer

Reply:
The concern there is, for inbound details (i.e. from proto) I have to carry around a converter as a private field inside the exception. And for outbound details (i.e. not yet to proto) I have to cast this type arg at runtime. So it starts to look something like:

```csharp
public T GetDetail<T>(int index = 0)
{
  // If there are inbound details use the converter, otherwise cast from outbound
  if (InboundDetails != null)
  {
    return converter!.ToValue<T>(InboundDetails[index]);
  }
  return (T) OutboundDetails[index];
}
```

I _guess_ that's ok so long as the user doesn't get the type wrong.

Reply:
Added this